### DISCUSSION TITLE
Trouble with many-to-many relationship Apollo/Nexus/Prisma

### DISCUSSION THREAD

**Discussion by jonhtyler** [prisma/prisma#11389] [Q&A] [answered]
Hi,  I have looked at the information at https://github.com/prisma/prisma/discussions/2429 and thought that was going to help me, but after continuing to have trouble for a few days, I am asking for help.

I am playing around with Apollo/Nexus/Prisma based on the tutorial located at https://www.howtographql.com/typescript-apollo/.  Most of what I have tried to do is working regarding adding/retrieving data from the database, when there is no relationship.  I am now at a point where I am trying to relate two tables/entities and I can't seem to get it to work.

This is my prisma schema:
```prisma
model Vehicle {
  id            Int             @id @unique @default(autoincrement())
  vin           String          @unique
  modelyear     Int
  make          String
  model         String
  owners        Owner[]
  VehicleOwners VehicleOwners[]
}

model Owner {
  id            Int             @id @unique @default(autoincrement())
  firstname     String
  lastname      String
  vehicleId     Int?
  vehicles      Vehicle[]
  VehicleOwners VehicleOwners[]
}

model VehicleOwners {
  vehicle    Vehicle  @relation(fields: [vehicleId], references: [id])
  vehicleId  Int
  owner      Owner    @relation(fields: [ownerId], references: [id])
  ownerId    Int
  assignedAt DateTime @default(now())

  @@id([vehicleId, ownerId])
}

```

And this is the beginning definition of my Owner.ts and Vehicle.ts files for the nexus configuration:

Vehicle.ts
```ts
export const Vehicle = objectType({
    name: "Vehicle",
    definition(t) {
        t.nonNull.id("id");
        t.nonNull.string("vin");
        t.nonNull.int("modelyear");
        t.nonNull.string("make");
        t.nonNull.string("model");
        t.list.nonNull.field("owners", {
            type: Owner,
            resolve: async (parent, args, context) => {
                let ownerList =  context.prisma.vehicleOwners.findMany({
                    where: {
                        vehicleId: {
                            equals: parent.id
                        }
                    },
                    include: {
                        owner: true,
                    }
                });

                console.log("Owners: " + JSON.stringify(ownerList))
                return ownerList;
            }
        })
    }
});

export const VehicleQuery = extendType({
    type: "Query",
    definition(t) {
        t.nonNull.list.field("getVehicles", {
            type: "Vehicle",
            resolve(parent, args, context, info) {
                return context.prisma.vehicle.findMany({
                    
                });
            }
        });
        t.field("getVehicleByVin", {
            type: "Vehicle",
            args: {
                vin: nonNull(stringArg())
            },
            resolve(parent, args, context, info) {
                var vehicle = context.prisma.vehicle.findFirst({
                    where: {
                        vin: args.vin
                    }
                }); 

                return vehicle;
            }
        })
    }
})

```

Owner.ts
```ts
export const Owner = objectType({
    name: "Owner",
    definition(t) {
        t.id("id");
        t.nonNull.string("firstname");
        t.nonNull.string("lastname");
        t.list.nonNull.field("vehicles", {
            type: Vehicle,
            resolve: async (parent, args, context) => {
                return context.prisma.vehicle.findMany({
                    where: {
                        id: parent.id
                    }
                });
            }
        })
    }
});

export const OwnersQuery = extendType({
    type: "Query",
    definition(t) {
        t.nonNull.list.nonNull.field("getOwners", {
            type: "Owner",
            resolve(parent, args, context, info) {
                return context.prisma.owner.findMany();
            },
        });
        t.field("getOwnerById", {
            type: "Owner",
            args: {
                id: nonNull(intArg())
            },
            resolve(parent, args, context, info) {
                var owner = context.prisma.owner.findFirst({
                    where: {
                        id: args.id
                    },
                });
                
                return owner;
            }
        })
    },
});

```

As you can see, in the Vehicle.ts code, I have tried to change the object type so that it will return the owners based on the information in the link I pasted above from another discussion on this repo.  That hasn't worked out at all.  Also, when I try this way, the line where I try to output the ownersList to the console comes back blank.  The owners table is not blank because I can query that table and get all the owners with the respective data returned as I had previously entered it.

Prior to that change, it looked similar to the Owner.ts file.  When I had it set up that way, I could get the Vehicle information, but the only owner information was based on the vehicleId that was returned with the parent property.  So, for instance, a vehicle with id of "1" would return the owner as owner with id of "1" as well, which is not correct.  

Can someone help me figure out the error of my ways? If there is a better place to post my question, please let me know that too.

**Comment by jonhtyler** [comment, marked as answer]
For what it is worth, I figured out what was going on with this.

The first thing I did was to simplify the Prisma schema to do an implicit relationship between vehicle and owner types and I removed the explicit many-to-many relationship definition.

```
model Owner {
  id        Int       @id @default(autoincrement())
  firstname String
  lastname  String
  email     String
  vehicles  Vehicle[]
}

model Vehicle {
  id        Int     @id @default(autoincrement())
  vin       String  @unique
  modelyear Int
  make      String
  model     String
  owners    Owner[]
}
```

The next thing I figured out is that I was creating the objectType incorrectly.  For example, in the above snippets, on the Owner.ts file, when I added the field for "vehicle," I was then querying the vehicle type (_context.prisma.vehicle.findMany()..._) which is why I was getting the owner with the same id as the vehicle.  It turns out that I needed to query the _owner_ type, which passes in the appropriate id for the owner, and then call the vehicles with that owner id:

```
export const Owner = objectType({
    name: "Owner",
    definition: t => {
        t.nonNull.int("id");
        t.nonNull.string("firstname");
        t.nonNull.string("lastname");
        t.nonNull.string("email");
        t.list.field("vehicle", { 
            type: "Vehicle",
            resolve: (parent, args, context, info) => {
                return context.prisma.owner.findUnique({
                    where: { id: parent.id }
                }).vehicles();
            }    
        })
    }
});
```

Once I saw how I had this wrong, and fixed it, everything started working as it should.

I also was having trouble when I switched to MS SQL server for my backing database.  When I started to code that, I couldn't figure out why the migration scripts appeared to work, and also data was being retrieved, but yet I couldn't see the data in SSMS console when I opened the database table list.  

It turns out that I was using "initial catalog" designation, which is common for SQL server connection strings.  However, the Prisma connection string uses the term "database" instead.  Because of this "bug" in my code, Prisma was applying the migration changes to the "master" database.  All of the tables and data created were in the master database instead of the database I created.  Just in case any one else runs into this problem.

Live and learn!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
